spec/jmap/push.mdown

```diff
@@ -1,69 +1,94 @@
 
-## Event Source
+    The client can compare the new state strings with its current values to see whether it has the current data for these types. If not, the changes can then be efficiently fetched in a single standard API request (using the *getFooUpdates* type methods).
+
+- **trigger**: `String`
+  What caused this change. The following causes are defined:
+
+    - `delivery`: The arrival of a new message caused the change.
+    - `user`: An action by the user caused the change.
+    - `unknown`: The cause of the change is unknown.
 
-There are two mechanisms by which the client can receive the push events. The first is directly via a `text/event-stream` resource, as described in
-<http://www.w3.org/TR/eventsource/>. This is essentially a long running HTTP request down which the server can push data. When a change occurs, the server MUST push an event called **state** to any connected clients.
+    Future specifications may define further values. Clients MUST treat an
+    unrecognised value the same as `unknown`.
 
-The server MAY also set a new `Last-Event-Id` that encodes the entire server state visible to the user. When a new connection is made to the event-source endpoint, the server can then work out whether the client has missed some changes which it should send immediately.
+    Clients in battery constrained environments may use this information to decide whether to immediately fetch the changes.
 
-The server MUST also send an event called **ping** with an empty object as the data if a maximum of 5 minutes has elapsed since the previous event. This MUST NOT set a new `Last-Event-Id`. A client may detect the absence of these to determine that the HTTP connection has been dropped somewhere along the route and so it needs to re-establish the connection.
 
-Refer to the Authentication section of this spec for details on how to get the URL for the event-source endpoint. The request must be authenticated using an `Authorization` header like any HTTP request.
+## PushSubscription
 
-A client MAY hold open multiple connections to the event-source, although it SHOULD try to use a single connection for efficiency.
+A push subscription is a message delivery context established between the client and a push service. A **PushSubscription** object has the following properties:
 
-## Web hook
+- **url**: `String`
+  An absolute URL where the JMAP server will POST the data for the push message.
+  This MUST begin with `https://`.
+- **expires**: `Date|null`
+  The time this push subscription expires. If specified, the JMAP server MUST NOT make further requests to this resource after this time. It MAY automatically remove the push subscription at or after this time.
+- **keys**: `Object|null`
+  Clients generated encryption keys. If supplied the server MUST use them as specified in <https://tools.ietf.org/html/draft-ietf-webpush-encryption-09> to encrypt all data sent to the push subscription. The object MUST have the following properties:
```

Comment:
I don't like that encryption is optional, but realize that it's only extra trouble if you are talking directly (as opposed to via a push service).  Can you describe exactly what conditions are necessary for the keys to be left unset?

Reply:
Well, basically if the push server is owned by the same entity as the JMAP server, you don't gain anything via the encryption and it just adds overhead. If it's going via a 3rd party then the client should use encryption. Something like that, wordsmithed a bit?